elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/filter/visitor/SplitFilterExpressionVisitor.java

```diff
@@ -0,0 +1,247 @@
+    public Pair<FilterExpression, FilterExpression> visitAndExpression(final AndFilterExpression expression) {
+        Pair<FilterExpression, FilterExpression> left = expression.getLeft().accept(this);
+        Pair<FilterExpression, FilterExpression> right = expression.getRight().accept(this);
+
+        if (isWhereExpression(left) && isWhereExpression(right)) {
+            // pure-W AND pure-W
+            return Pair.of(
+                    new AndFilterExpression(
+                            extractWhereExpression(left),
+                            extractWhereExpression(right)
+                    ),
+                    null
+            );
+        } else if (isHavingExpression(left) && isWhereExpression(right)) {
+            // pure-H AND pure-W
+            return Pair.of(
+                    extractHavingExpression(left),
+                    extractWhereExpression(right)
+            );
+
+        } else if (isWhereExpression(left) && isHavingExpression(right)) {
+            // pure-W AND pure-H
+            return Pair.of(
+                    extractWhereExpression(left),
+                    extractHavingExpression(right)
+            );
+        } else {
+            return Pair.of(
+                    AndFilterExpression.withLeftAndRight(
+                            extractWhereExpression(left),
+                            extractWhereExpression(right)
+                    ),
+                    AndFilterExpression.withLeftAndRight(
+                            extractHavingExpression(left),
+                            extractHavingExpression(right)
+                    )
+            );
+        }
+    }
+
+    @Override
+    public Pair<FilterExpression, FilterExpression> visitOrExpression(final OrFilterExpression expression) {
+        Pair<FilterExpression, FilterExpression> left = expression.getLeft().accept(this);
+        Pair<FilterExpression, FilterExpression> right = expression.getRight().accept(this);
+
+        if (isWhereExpression(left) && isWhereExpression(right)) {
+            // W OR W
+            return Pair.of(
+                    new OrFilterExpression(
+                            extractWhereExpression(left),
+                            extractWhereExpression(right)
+                    ),
+                    null
+            );
+        } else {
+            // H OR H
+            // W OR H (W promote)
+            // H OR W (W promote)
+            return Pair.of(
+                    null,
+                    new OrFilterExpression(
+                            extractHavingExpression(left),
+                            extractHavingExpression(right)
+                    )
+            );
+        }    }
+
+    @Override
+    public Pair<FilterExpression, FilterExpression> visitNotExpression(NotFilterExpression expression) {
+        FilterExpression normalized = getNormalizationVisitor().visitNotExpression(expression);
+
+        if (normalized instanceof AndFilterExpression) {
```

Comment:
This might be missing the not expression.  Normalizing a not can still return a not.